src/tasks/TaskHelper.ts

```diff
@@ -0,0 +1,166 @@
+        tasks.registerTaskProvider(
+            'docker-build',
+            new DockerBuildTaskProvider(helpers)
+        )
+    );
+
+    ctx.subscriptions.push(
+        tasks.registerTaskProvider(
+            'docker-run',
+            new DockerRunTaskProvider(helpers)
+        )
+    );
+}
+
+export async function addTask(task: DockerBuildTaskDefinition | DockerRunTaskDefinition): Promise<boolean> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    if (allTasks.some(t => t.label === task.label)) {
+        return false;
+    }
+
+    allTasks.push(task);
+    await workspaceTasks.update('tasks', allTasks);
+    return true;
+}
+
+export async function getAssociatedDockerRunTask(debugConfiguration: DockerDebugConfiguration): Promise<DockerRunTaskDefinition | undefined> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks: TaskDefinition[] = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    return await recursiveFindTaskByType(allTasks, 'docker-run', debugConfiguration);
+}
+
+export async function getAssociatedDockerBuildTask(runTask: DockerRunTaskDefinition): Promise<DockerBuildTaskDefinition | undefined> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks: TaskDefinition[] = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    return await recursiveFindTaskByType(allTasks, 'docker-build', runTask);
+}
+
+export async function getOfficialBuildTaskForDockerfile(dockerfile: string, folder: WorkspaceFolder): Promise<Task | undefined> {
+    let buildTasks = await tasks.fetchTasks({ type: 'docker-build' });
+    buildTasks =
+        buildTasks.filter(t => t.execution.args.some(a => { // Find all build tasks where an argument to 'docker build' is this Dockerfile
+            let arg: string;
+            if (typeof a === 'string') {
+                arg = a;
+            } else {
+                arg = a.value;
+            }
+
+            arg = resolveWorkspaceFolderPath(folder, arg);
+            return arg === dockerfile;
+        }));
+
+    if (buildTasks.length === 1) {
+        return buildTasks[0]; // If there's only one build task, take it
+    } else if (buildTasks.length > 1) {
+        return buildTasks.find(t => t.name === 'docker-build: release') || buildTasks[0]; // If there's multiple try finding one with the name 'docker-build: release', else take first

```

Comment:
>docker-build: release [](start = 48, length = 21)

We shouldn't take a dependency on the name of the task.

Reply:
Do you have an alternative suggestion? Adding another field just for this is overly verbose.

Reply:
Here's my suggestion: If there's more than one, and one is called "docker-build: release" then use it (the most common case); otherwise quickpick.

Reply:
Early on we had discussed adding a "config is for debugging" flag in the definition to allow us to more easily locate the likeliest configuration (as there's likely only two configurations per Dockerfile).  The Node.js `TaskHelper` already has such a flag (as it's needed for other reasons).  We could either make that a common flag or defer selection to the individual task helpers (which could use their own, potentially different logic).

I think I prefer the first option (the common debugging flag) myself.

Reply:
I don't agree with that approach. Allow me to explain.

If we add such a flag, we cannot use it to control logic within the task helpers. To do so would be to do "magic", something the users cannot look into (unless they go digging through our source), and they cannot control, etc. Magic === evil.

So that would leave the flag only useful for this one specific scenario of deciding which build config is the release one vs. debug one. If we think that users might change the name and take "release" out, they can just as easily change this hypothetical flag. If we fall back to using a quick-pick if there's no matching config of this name, what's the down side? We avoid an unnecessary flag and have equally functional behavior.

Reply:
How would you feel about an 'isDefaultBuildTask' flag to indicate the task to use by default, with a fallback to quick pick when no default (or multiple defaults) are found?

Reply:
VSCode has a notion for that already; you can see it with Ctrl+Shift+P > Configure default build task. It adds `"group": { "kind": "build", "isDefault" : true }`. I wouldn't really want to add something that has similar appearance but totally different effect.

Reply:
What if we called it "isDefaultDockerTask"?

Reply:
Yeah, that would be better. I'm going to think about it some and see if there's a clear, existing indicator. I still think a dedicated property may be a bit over-engineered, especially considering as far as we know this debug/release build scenario only really applies to .NET Core.